README.md
# Assam Tribune Archive v0.2

The [Assam Tribune](http://assamtribune.com) is the largest English language daily of North-East India. But the website is prehistoric with no proper archival system.

v0.2 script scrapes images of Assam Tribune daily and converts in into a single PDF for archival.

Very crude code. Patches are welcome. 

# Installation

    pip install --upgrade -r requirements.txt
    python3 src/atscraper.py

Set up Chron job to run everyday.

# To be implemented in the future if time permits

* Optimize the file size of the PDF
* Integration with mailchimp for daily inbox delivery
* Downloading old editions of the paper
* OCR the images for a search database




changes.txt
0.1

* Initial commit

0.2

* Complete rewrite of the script to modularize the code into functions


test.py
import grequests as async
# If using requests > v0.13.0, use
# from grequests import async

urls = [
    'http://python-requests.org',
    'http://httpbin.org',
    'http://python-guide.org',
    'http://kennethreitz.com'
]

urls2 = []

# A simple task to do to each response object
def do_something(response):
    urls2.append = print(response.url)

# A list to hold our things to do via async
async_list = []

for u in urls:
    # The "hooks = {..." part is where you define what you want to do
    # 
    # Note the lack of parentheses following do_something, this is
    # because the response will be used as the first argument automatically
    action_item = async.get(u, hooks = {'response' : do_something})

    # Add the task to our list of things to do via async
    async_list.append(action_item)

# Do our list of things to do via async
async.map(async_list)


src/atscraper.py
#!/usr/bin/python3

from bs4 import BeautifulSoup #For parsing the website
import requests 
import os 
import glob #For processing the downloaded JPG files 
from fpdf import FPDF #For working with PDFs
from datetime import datetime #For naming the exported PDF file
import pytz #For accounting for Indian Time Zone in the file name
import re
import sys

def get_todays_homepage_url(url):
    """Gives the homepage URL to today's e-paper
    """
    r = requests.get(url)
    c = r.content #HTML content
    soup = BeautifulSoup(c,'lxml') #Making HTML content legible
    soup.find_all('meta')
    content = soup.select_one("meta[content*=url]")["content"]
    url = content.split(";")[-1].split("url=")[-1]
    sp = url.split('?') #Splittin the URL to what I need
    dailyurl = '/'.join(sp[0].split('/')[:4])+'/at.asp?'+sp[1]+'/'
    return dailyurl

def natural_sort(l):
    """Sorts by natural order of numbers
    """
    convert = lambda text: int(text) if text.isdigit() else text.lower() 
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] 
    return sorted(l, key = alphanum_key)

def get_all_image_links(dailyurl):
    """Collates URLs of all images
    """
    r2 = requests.get(dailyurl)
    c2 = r2.content
    soup2 = BeautifulSoup(c2, 'lxml')
    #Storing all the Page URLs
    pageLinks = []
    imageLinks = []
    for anchor in soup2.find_all('a', {"class" : "PTopLink"}, href=True):
        pageLinks.append('http://www.assamtribune.com/scripts/'+ anchor['href'])
        imageLinks.append(os.path.join('http://www.assamtribune.com/scripts/spat.'+
        re.split('/|\.|=', anchor['href'])[1] + "=" + datetime.now().strftime("%Y"),
        re.split('/|\.|=', anchor['href'])[2], 
        'Big'+re.split('/|\.|=', anchor['href'])[3]+'.jpg'))        
    return imageLinks


def download_all_images(imageLinks):
    """Downloads all the images in full res
    """
    for image in imageLinks:
        r3 =requests.get(image,stream=True)
        soup3 = BeautifulSoup(r3.text,"lxml")
        img_link = soup3.find_all("img")
        download_url = "http://www.assamtribune.com"
        for link in img_link:
            download_url+=link["src"][2:]
            break
        r4 = requests.get(download_url, stream=True)
        if r4.status_code == 200:
            with open(os.path.normpath(os.getcwd() + os.sep + os.pardir) + "/img/" + re.findall('\d+', image)[2]+".jpg", 'wb') as f:
                f.write(r4.content)
	
def convert_images_to_pdf(imagelist, pdftitle, pdfcreator):
    """Convert all images to PDF. 
    imagelist is the list with all image filenames.
    """
    pdf = FPDF()
    for image in imagelist:
        pdf.add_page()
        pdf.image(image, w=170)
        pdf.set_title(pdftitle)
        pdf.set_creator(pdfcreator)
    at_filedate = datetime.now(pytz.timezone('Asia/Kolkata')).strftime("%Y%m%d")
    suffix = 'pdf'    
    at_filename = os.path.join(os.path.normpath(os.getcwd() + os.sep + os.pardir), 'pdf', at_filedate + "_assamtribune" + os.extsep + suffix)
    pdf.output(at_filename, "F")


def main():
    """One function to rule them all.
    """
    y = get_todays_homepage_url("http://www.assamtribune.com/")
    z = get_all_image_links(y)
    download_all_images(z)
    imagelist = natural_sort(glob.glob('../img/*.jpg'))
    convert_images_to_pdf(imagelist, 'Assam Tribune Daily', 'Manabendra Saharia')

if __name__ == '__main__':
    sys.exit(main())


src/atscrape2pdf.py
#!/usr/bin/python3

from bs4 import BeautifulSoup #For parsing the website
import requests 
import os 
import glob #For processing the downloaded JPG files 
from fpdf import FPDF #For working with PDFs
from datetime import datetime #For naming the exported PDF file
import pytz #For accounting for Indian Time Zone in the file name
import re

r = requests.get('http://www.assamtribune.com/') #Page Response
c = r.content #HTML content
soup = BeautifulSoup(c,'lxml') #Making HTML content legible

soup.find_all('meta')
content = soup.select_one("meta[content*=url]")["content"]
url = content.split(";")[-1].split("url=")[-1]
 
sp = url.split('?')

dailyurl = '/'.join(sp[0].split('/')[:4])+'/at.asp?'+sp[1]+'/'

r2 = requests.get(dailyurl)
c2 = r2.content
soup2 = BeautifulSoup(c2,'lxml')

#Storing all the Page URLs
pageLinks = []
imageLinks = []
imageLinks2 = []

for anchor in soup2.find_all('a', { "class" : "PTopLink" }, href=True): #Storing all URLs from HTML
     pageLinks.append('http://www.assamtribune.com/scripts/'+ anchor['href'])  
     imageLinks.append(os.path.join('http://www.assamtribune.com/scripts/spat.'+
                 re.split('/|\.|=', anchor['href'])[1] + "=" + datetime.now().strftime("%Y") ,
                 re.split('/|\.|=', anchor['href'])[2] , 
                 'Big'+re.split('/|\.|=', anchor['href'])[3]+'.jpg'))

for image in imageLinks:
    r3 =requests.get(image,stream=True)
    soup3 = BeautifulSoup(r3.text,"lxml")
    img_link = soup3.find_all("img")
    download_url = "http://www.assamtribune.com"
    for link in img_link:
        download_url+=link["src"][2:]
        break
    r4 = requests.get(download_url, stream=True)
    if r4.status_code == 200:
        #with open(os.getcwd() + "/img/" + image.split("/")[-1], 'wb') as f:
        with open(os.path.normpath(os.getcwd() + os.sep + os.pardir) + "/img/" + re.findall('\d+', image)[2]+".jpg", 'wb') as f:
            f.write(r4.content)

#For sorting the pages by natural order

def natural_sort(l): 
    convert = lambda text: int(text) if text.isdigit() else text.lower() 
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] 
    return sorted(l, key = alphanum_key)

imagelist = natural_sort(glob.glob('../img/*.jpg'))

#Coverting the images to PDF
pdf = FPDF()
# imagelist is the list with all image filenames
for image in imagelist:
    pdf.add_page()
    pdf.image(image, w=170)
    pdf.set_title('Assam Tribune Daily')
    pdf.set_creator('Manabendra Saharia')

at_filedate = datetime.now(pytz.timezone('Asia/Kolkata')).strftime("%Y%m%d")
suffix = 'pdf'    
at_filename = os.path.join(os.path.normpath(os.getcwd() + os.sep + os.pardir), 'pdf', at_filedate + "_assamtribune" + os.extsep + suffix)

pdf.output(at_filename, "F")


src/atscrape2pdf2.py
#!/usr/bin/env python

from bs4 import BeautifulSoup #For parsing the website
import requests 
import os 
import glob #For processing the downloaded JPG files 
from fpdf import FPDF #For working with PDFs
from datetime import datetime #For naming the exported PDF file
import pytz #For accounting for Indian Time Zone in the file name
import re
import erequests #For Asyncronous downloading of images

r = requests.get('http://www.assamtribune.com/') #Page Response
c = r.content #HTML content
soup = BeautifulSoup(c,'lxml') #Making HTML content legible

soup.find_all('meta')
content = soup.select_one("meta[content*=url]")["content"]
url = content.split(";")[-1].split("url=")[-1]
 
sp = url.split('?')

dailyurl = '/'.join(sp[0].split('/')[:4])+'/at.asp?'+sp[1]+'/'

r2 = requests.get(dailyurl)
c2 = r2.content
soup2 = BeautifulSoup(c2,'lxml')

#Storing all the Page URLs
pageLinks = []
imageLinks = []
imageLinks2 = []

for anchor in soup2.find_all('a', { "class" : "PTopLink" }, href=True): #Storing all URLs from HTML
     pageLinks.append('http://www.assamtribune.com/scripts/'+ anchor['href'])  
     imageLinks.append(os.path.join('http://www.assamtribune.com/scripts/spat.'+
                 re.split('/|\.|=', anchor['href'])[1] + "=" + datetime.now().strftime("%Y") ,
                 re.split('/|\.|=', anchor['href'])[2] , 
                 'Big'+re.split('/|\.|=', anchor['href'])[3]+'.jpg'))

rs = [erequests.async.get(u) for u in imageLinks]
list(erequests.map(rs))

# erequests.map will call each async request to the action
# what returns processed request `req`
for req in erequests.map(rs):
    if req.ok:
        content = req.content
        # process it here
        print(req.url)

#for image in imageLinks:
#    r3 =requests.get(image,stream=True)
#    soup3 = BeautifulSoup(r3.text,"lxml")
#    img_link = soup3.find_all("img")
#    download_url = "http://www.assamtribune.com"
#    for link in img_link:
#        download_url+=link["src"][2:]
#        break
#    r4 = requests.get(download_url, stream=True)
#    if r4.status_code == 200:
#        #with open(os.getcwd() + "/img/" + image.split("/")[-1], 'wb') as f:
#        with open(os.getcwd() + "/img/" + re.findall('\d+', image)[2]+".jpg", 'wb') as f:
#            f.write(r4.content)

#For sorting the pages by natural order

def natural_sort(l): 
    convert = lambda text: int(text) if text.isdigit() else text.lower() 
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] 
    return sorted(l, key = alphanum_key)

imagelist = natural_sort(glob.glob('img/*.jpg'))

#Coverting the images to PDF
pdf = FPDF()
# imagelist is the list with all image filenames
for image in imagelist:
    pdf.add_page()
    pdf.image(image, w=170)
    pdf.set_title('Assam Tribune Daily')
    pdf.set_creator('Manabendra Saharia')

at_filedate = datetime.now(pytz.timezone('Asia/Kolkata')).strftime("%Y%m%d")
suffix = 'pdf'    
at_filename = os.path.join(os.getcwd(), 'pdf', at_filedate + "_assamtribune" + os.extsep + suffix)

pdf.output(at_filename, "F")


src/atscraper_dev.py
#!/usr/bin/python3

from bs4 import BeautifulSoup #For parsing the website
import requests 
import os 
import glob #For processing the downloaded JPG files 
from fpdf import FPDF #For working with PDFs
from datetime import datetime #For naming the exported PDF file
import pytz #For accounting for Indian Time Zone in the file name
import re
import sys

#Steps for this script
#1 Use requests and beautiful soup to get the links of images
#2 Download the images
#3 Convert to one PDF
#4 Write out the file

def get_todays_homepage_url(url):
    """Gives the homepage URL to today's e-paper
    """
    r = requests.get(url)
    c = r.content #HTML content
    soup = BeautifulSoup(c,'lxml') #Making HTML content legible
    soup.find_all('meta')
    content = soup.select_one("meta[content*=url]")["content"]
    url = content.split(";")[-1].split("url=")[-1]
    sp = url.split('?') #Splittin the URL to what I need
    dailyurl = '/'.join(sp[0].split('/')[:4])+'/at.asp?'+sp[1]+'/'
    return dailyurl

def natural_sort(l):
    """Sorts by natural order of numbers
    """
    convert = lambda text: int(text) if text.isdigit() else text.lower() 
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] 
    return sorted(l, key = alphanum_key)

def get_all_image_links(dailyurl):
    """Collates URLs of all images
    """
    r2 = requests.get(dailyurl)
    c2 = r2.content
    soup2 = BeautifulSoup(c2, 'lxml')
    #Storing all the Page URLs
    pageLinks = []
    imageLinks = []
    for anchor in soup2.find_all('a', {"class" : "PTopLink"}, href=True):
        pageLinks.append('http://www.assamtribune.com/scripts/'+ anchor['href'])
        imageLinks.append(os.path.join('http://www.assamtribune.com/scripts/spat.'+
        re.split('/|\.|=', anchor['href'])[1] + "=" + datetime.now().strftime("%Y"),
        re.split('/|\.|=', anchor['href'])[2], 
        'Big'+re.split('/|\.|=', anchor['href'])[3]+'.jpg'))        
    return imageLinks

def download_all_images(imageLinks):
    """Downloads all the images in full res
    """
    for image in imageLinks:
        r3 =requests.get(image,stream=True)
        soup3 = BeautifulSoup(r3.text,"lxml")
        img_link = soup3.find_all("img")
        download_url = "http://www.assamtribune.com"
        for link in img_link:
            download_url+=link["src"][2:]
            break
        r4 = requests.get(download_url, stream=True)
        if r4.status_code == 200:
            with open(os.path.normpath(os.getcwd() + os.sep + os.pardir) + "/img/" + re.findall('\d+', image)[2]+".jpg", 'wb') as f:
                f.write(r4.content)

def convert_images_to_pdf(imagelist, pdftitle, pdfcreator):
    """Convert all images to PDF. 
    imagelist is the list with all image filenames.
    """
    pdf = FPDF()
    for image in imagelist:
        pdf.add_page()
        pdf.image(image, w=170)
        pdf.set_title(pdftitle)
        pdf.set_creator(pdfcreator)
    at_filedate = datetime.now(pytz.timezone('Asia/Kolkata')).strftime("%Y%m%d")
    suffix = 'pdf'    
    at_filename = os.path.join(os.path.normpath(os.getcwd() + os.sep + os.pardir), 'pdf', at_filedate + "_assamtribune" + os.extsep + suffix)
    pdf.output(at_filename, "F")


def main():
    """One function to rule them all.
    """
    y = get_todays_homepage_url("http://www.assamtribune.com/")
    z = get_all_image_links(y)
    download_all_images(z)
    imagelist = natural_sort(glob.glob('../img/*.jpg'))
    convert_images_to_pdf(imagelist, 'Assam Tribune Daily', 'Manabendra Saharia')

if __name__ == '__main__':
    sys.exit(main())
